google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java

```diff
@@ -0,0 +1,428 @@
+          // Getting the result of the SettableApiFuture that contains the multiplexed session will
+          // also automatically propagate any error that happened during the creation of the
+          // session, such as for example a DatabaseNotFound exception. We therefore do not need
+          // any special handling of such errors.
+          multiplexedSessionReference.get().get(),
+          singleUse ? getSingleUseChannelHint() : NO_CHANNEL_HINT,
+          singleUse);
+    } catch (ExecutionException executionException) {
+      throw SpannerExceptionFactory.asSpannerException(executionException.getCause());
+    } catch (InterruptedException interruptedException) {
+      throw SpannerExceptionFactory.propagateInterrupt(interruptedException);
+    }
+  }
+
+  private DelayedMultiplexedSessionTransaction createDelayedMultiplexSessionTransaction() {
+    return new DelayedMultiplexedSessionTransaction(
+        this, tracer.getCurrentSpan(), multiplexedSessionReference.get());
+  }
+
+  private int getSingleUseChannelHint() {
+    if (this.numCurrentSingleUseTransactions.incrementAndGet() > this.numChannels) {
+      return NO_CHANNEL_HINT;
+    }
+    synchronized (this.channelUsage) {
+      // Get the first unused channel.
+      int channel = this.channelUsage.nextClearBit(/* fromIndex = */ 0);
+      // BitSet returns an index larger than its original size if all the bits are set.
+      // This then means that all channels have already been assigned to single-use transactions,
+      // and that we should not use a specific channel, but rather pick a random one.
+      if (channel == this.numChannels) {
+        return NO_CHANNEL_HINT;
+      }
+      this.channelUsage.set(channel);
+      return channel;
+    }
+  }
+
+  @Override
+  public ReadContext singleUse() {
+    return createMultiplexedSessionTransaction(true).singleUse();
+  }
+
+  @Override
+  public ReadContext singleUse(TimestampBound bound) {
+    return createMultiplexedSessionTransaction(true).singleUse(bound);
+  }
+
+  @Override
+  public ReadOnlyTransaction singleUseReadOnlyTransaction() {
+    return createMultiplexedSessionTransaction(true).singleUseReadOnlyTransaction();
+  }
+
+  @Override
+  public ReadOnlyTransaction singleUseReadOnlyTransaction(TimestampBound bound) {
+    return createMultiplexedSessionTransaction(true).singleUseReadOnlyTransaction(bound);
+  }
+
+  @Override
+  public ReadOnlyTransaction readOnlyTransaction() {
+    return createMultiplexedSessionTransaction(false).readOnlyTransaction();
+  }
+
+  @Override
+  public ReadOnlyTransaction readOnlyTransaction(TimestampBound bound) {
+    return createMultiplexedSessionTransaction(false).readOnlyTransaction(bound);
+  }
+
+  /**
+   * It is enough with one executor to maintain the multiplexed sessions in all the clients, as they
+   * do not need to be updated often, and the maintenance task is light.
+   */
+  private static final ScheduledExecutorService MAINTAINER_SERVICE =
```

Comment:
It would have been slightly harder, but that was not my main reason for not using the same thread. My reasoning for creating a separate maintainer and executor was:
1. This separates multiplexed sessions completely from the session pool. That means that once all operations work on multiplexed sessions, we can just remove the entire session pool implementation.
2. This maintainer uses a `ScheduledThreadPoolExecutor` with zero core threads, and only runs the task once every 10 minutes. The additional resource usage is therefore minimal.